done hw05

cepmap opened the pull request:
## Домашнее задание №5 «Параллельное исполнение»

### Чек-лист студента ([Что это?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0))
- [x] Я перечитал задание после решения. [Зачем перечитывать задание?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%82%D1%8B%D0%B2%D0%B0%D1%82%D1%8C-%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5)
- [x] Я запустил `go mod tidy`.
- [x] Я удалил `.sync` файл. [Зачем его удалять?](https://github.com/OtusGolang/home_work/wiki/%5B%D0%A1%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0%D0%BC%5D-%D0%9F%D1%80%D0%BE%D1%86%D0%B5%D1%81%D1%81-%D1%81%D0%B4%D0%B0%D1%87%D0%B8-%D0%94%D0%97#user-content-%D0%92%D0%B0%D1%80%D0%B8%D0%B0%D0%BD%D1%82-2)
- [x] Я сравнил объем кода с объемом авторского решения. [Зачем сверять объем кода?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D1%81%D0%B2%D0%B5%D1%80%D1%8F%D1%82%D1%8C-%D0%BE%D0%B1%D1%8A%D0%B5%D0%BC-%D0%BA%D0%BE%D0%B4%D0%B0)
  - run.go ~55 строк

### Критерии оценки
- [x] Пайплайн зелёный - 4 балла
- [x] Добавлены новые юнит-тесты - до 4 баллов
- [x] Понятность и чистота кода - до 2 баллов

#### Зачёт от 7 баллов

cepmap committed:
done hw05

cepmap committed:
Merge branch 'master' into hw05_parallel_execution

hungry-honeybadger commented on hw05_parallel_execution/run.go:
```diff
@@ -2,14 +2,63 @@ package hw05parallelexecution
+	}
+
+	wg.Wait()
+	if errCounter.Get() < n+m {
```
Ошибку мы должно вернуть если число ошибок больши или равно m. n+m это максимальное число горутин которое должно отработать

hungry-honeybadger commented on hw05_parallel_execution/run.go:
```diff
@@ -2,14 +2,63 @@ package hw05parallelexecution
+
+	wg.Wait()
+	if errCounter.Get() < n+m {
+		ErrErrorsLimitExceeded = nil
```
Подумайте как отработает код если я 2 раза подряд вызову метод Run. Если при первом вызове сработает это условие, то при 2 условие код вообще сможет вернуть эту ошибки или мы ее теряем

hungry-honeybadger commented on hw05_parallel_execution/run.go:
```diff
@@ -2,14 +2,63 @@ package hw05parallelexecution
+		m = 0
+	}
+	errCounter := errCounter{}
+	tasksChannel := make(chan Task, len(tasks))
```
В данном случае использование буферезированного канала неизвестного размера не оправдано. Это и лишние расходы памяти и трата времени на копирование тасков из слайса в канал. По сути его можно заменить потокобезопасный index для слайса и такой расклад будет работать лучше. Поэтому за такое решение поставить более 8 балов не смогу, в идеале тут должен быть не буферезированый канала

cepmap committed:
1. Был кривоватый тест, написанный в 2 часа ночи. Поправлено.
2. Поправил замечания по поводу проверки счётчика ошибок и переопределения ошибки в nil.

cepmap committed:
Merge remote-tracking branch 'origin/hw05_parallel_execution' into hw05_parallel_execution

cepmap committed:
fix int32 to int 32

cepmap committed:
fixed comment via unbuffered channel
